Simplify operating on sets and maps of dialect ops.

tsymalla opened the pull request:
This commit introduces a new type, `OpSet`, that simplifies membership tests of an instruction on a set of dialect operations. With this PR, applying a callback on a given set of operations also becomes simpler since a visitor function `addSet` is introduced.

It also introduces a map-like data structure, `OpMap`, which can be used to simplify handling of association of dialect operations to certain values.

With this PR, a new check-llvm-dialects-units target is added which tests the new GTest ADT tests as well.

nhaehnle commented on example/ExampleMain.cpp:
```diff
@@ -4,28 +4,31 @@
+ *  of this software and associated documentation files (the "Software"), to
+ *deal in the Software without restriction, including without limitation the
+ *rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
+ *sell copies of the Software, and to permit persons to whom the Software is
```
Unrelated changes.

nhaehnle commented on example/ExampleMain.cpp:
```diff
@@ -166,12 +172,33 @@ LLVM_DIALECTS_VISITOR_PAYLOAD_PROJECT_FIELD(VisitorNest, inner)
+              } else if (auto *II = dyn_cast<IntrinsicInst>(&op)) {
+                if (II->getIntrinsicID() == Intrinsic::umin) {
+                  *self.out << "visiting umin (set): " << op << '\n';
+                }
```
Rephrase the checks here as assertions, since the use of `complexSet` should make these a guarantee. We don't want to accidentally miss a bug here.

nhaehnle commented on include/llvm-dialects/Dialect/Dialect.h:
```diff
@@ -249,7 +249,9 @@ bool isSimpleOperationDecl(const llvm::Function *fn, llvm::StringRef name);
 bool isSimpleOperation(const llvm::CallInst *i, llvm::StringRef name);
+bool isSimpleOperation(llvm::StringRef fn, llvm::StringRef name);
 bool isOverloadedOperation(const llvm::CallInst *i, llvm::StringRef name);
+bool isOverloadedOperation(llvm::StringRef fn, llvm::StringRef name);
```
I don't think you need these overloads. Instead, add an `isOperationDecl` (see above) that takes an `isOverloaded` argument in addition to the name.

nhaehnle commented on include/llvm-dialects/Dialect/Utils.h:
```diff
@@ -78,4 +78,7 @@ template <typename T> Printable<T> printable(const T &x) {
 }
 
+bool isMatchingFunc(llvm::StringRef refFn, bool isOverload,
+                    llvm::StringRef mnemonic);
```
See the other comment -- I don't think this is needed, just an `isOperationDecl` is fine and fits better to what we already have.

nhaehnle commented on include/llvm-dialects/Dialect/OpMap.h:
```diff
@@ -0,0 +1,356 @@
+           m_intrinsics == rhs.m_intrinsics && m_dialectOps == rhs.m_dialectOps;
+  }
+
+  bool operator!=(const OpMap &rhs) const { return !(*this == rhs); }
```
I don't think these `operator==/!=` implement the correct semantics, since they depend on the ordering of the dialect ops vector. Do we need them in the first place?

tsymalla replied:
Right now, unfortunately yes, otherwise the comparison tests don't work (see the `OpMapIRTests`).
You are right on the dependency of the ordering. Maybe it makes sense to always sort the vector once it was mutated.

tsymalla replied:
I just went with the approach to do a quick pairwise lookup, which should work fine.

nhaehnle commented on include/llvm-dialects/Dialect/OpMap.h:
```diff
@@ -0,0 +1,356 @@
+  // -------------------------------------------------------------
+
+  // Check if the map contains either a core operation or an intrinsic.
+  bool contains(unsigned op) const {
```
This needs to be split into two methods (or removed), because a core op code can alias an intrinsic ID.

nhaehnle commented on include/llvm-dialects/Dialect/OpMap.h:
```diff
@@ -0,0 +1,356 @@
+      for (const unsigned op : desc.getOpcodes()) {
+        if (m_coreOpcodes.contains(op))
+          return m_coreOpcodes[op];
+      }
```
I think this loop, which has a dependency on the hash function used in `DenseMap`, proves that the `OpMap` class should refuse `OpDescription`s of multiple core opcodes and intrinsics in most or even all places.

This would also allow you to directly forward this to the correct `DenseMap` instead of forwarding to `insert_or_get` where a redundant kind check happens (which the compiler can probably optimize away if it inlines everything, but still).

nhaehnle commented on include/llvm-dialects/Dialect/OpMap.h:
```diff
@@ -0,0 +1,356 @@
+  // -------------------------------------------------------------
+  ValueT &operator[](const IntrinsicInst &inst) {
+    return m_intrinsics[inst.getIntrinsicID()];
+  }
```
It's weirdly non-orthogonal to have this overload but not one that works for general instructions. I know that we just can't have the latter, but I think this means we shouldn't have this overload either.

tsymalla replied:
Makes sense, thanks.

nhaehnle commented on include/llvm-dialects/Dialect/OpMap.h:
```diff
@@ -0,0 +1,356 @@
+    return lookup(OpDescription::get<OpT>(), func);
+  }
+
+  ValueT lookup(const OpDescription &desc, const Function &func) const {
```
What is this and similar overloads good for? If we already had an `OpDescription`, then we could just lookup based on that, couldn't we? I suspect they should be removed.

nhaehnle commented on include/llvm-dialects/Dialect/OpMap.h:
```diff
@@ -0,0 +1,356 @@
+  // a dialect operation or a core instruction.
+  ValueT lookup(const Instruction &inst) const {
+    if (isa<IntrinsicInst>(inst) || isa<CallInst>(inst))
+      return lookup(*cast<CallInst>(inst).getCalledFunction());
```
This can just be `isa<CallInst>`, since `IntrinsicInst` is derived from `CallInst`.

I wonder whether this case should fallback to looking up `OpDescription::get<CallInst>()`. I tend to think that it should, though in any case it feels likely that users of this functionality will always need some amount of special-casing around `CallInst`.

Note that this also needs to handle the case that `getCalledFunction` returns nullptr.

nhaehnle commented on include/llvm-dialects/Dialect/OpMap.h:
```diff
@@ -0,0 +1,356 @@
+        erased |= m_intrinsics.erase(op);
+    }
+
+    // Try to erase the dialect op at last.
```
This should be under an `else` (or just always directly `return` from the cases above)

nhaehnle commented on test/unit/dialect/TestDialect.td:
```diff
@@ -0,0 +1,73 @@
+  let results = (outs);
+  let arguments = (ins);
+
+  let defaultBuilderHasExplicitResultType = true;
```
That doesn't make a lot of sense when there's no result value in the first place :)

I think this test dialect needs at least two ops, one with and one without overloads.

tsymalla replied:
Hm, the `DialectOp4` is used to show the intended behavior when a function is overloaded. Here, the same function is used with an `i32` and a `float` argument. This can be seen in the `DialectOpOverloadTests` - but I can extend the testing, if you want.

nhaehnle commented on include/llvm-dialects/Dialect/OpSet.h:
```diff
@@ -0,0 +1,267 @@
+  bool matches(const OpDescription &desc, const Instruction &inst) const {
+    if (isa<IntrinsicInst>(inst) || isa<CallInst>(inst)) {
+      return matches(desc, *cast<CallInst>(inst).getCalledFunction());
+    }
```
I notice there aren't any unit tests for `matches`. And similar to those analogous `lookup` overloads in `OpMap`, I don't understand what it's good for.

nhaehnle commented on include/llvm-dialects/Dialect/OpSet.h:
```diff
@@ -0,0 +1,267 @@
+  }
+
+  // Checks if `inst` does not belong to the OpSet.
+  bool isNoMember(const Instruction &inst) const { return !isMember(inst); }
```
Why?

nhaehnle commented on include/llvm-dialects/Dialect/OpSet.h:
```diff
@@ -0,0 +1,267 @@
+  // -------------------------------------------------------------
+
+  // Checks if `inst` belongs to the OpSet.
+  bool isMember(const Instruction &Inst) const {
```
Why not make this an overload of `contains`. That feels a bit more intuitive to me.

nhaehnle commented on include/llvm-dialects/Dialect/OpSet.h:
```diff
@@ -0,0 +1,267 @@
+  }
+
+  // Checks if `func` does not belong to the OpSet.
+  bool isNoMember(const Function &func) const { return !isMember(func); }
```
Why?

nhaehnle reviewed (commented):
Thank you, I like that you were quite thorough with the interface of OpSet and OpMap. I think you went a bit overboard in some places, but comments are inline.

I still need to take a closer look at the visitor parts.

tsymalla commented:
Regenerated copyright comments.
Interestingly, local clang-format (16.0.6) breaks the comments for me.

tsymalla commented:
Side note - I did not work on all comments yet, just the obvious ones.

nhaehnle reviewed (commented):
The intention for the visitor implementation was to merge the set into the existing data structures, or perhaps just replace the existing data structures by an `OpMap<SmallVector<unsigned>>`. This is both for cleanliness and for performance reasons.

tsymalla commented:
Makes sense, will take care of it.

tsymalla commented:
Update to address the remaining change requests.

I thought about the Visitor changes a bit more. Right now, in my opinion it only makes sense to either use the `OpMap` approach or to handle the set cases separately.
The issue with the current Visitor implementation is that for `OpDescription` objects, it relies on having a single instance obtained by OpDescription::get<OpT>(), which is static in the generated dialect. This works since for each `OpT`, there is a separate instance in the according `::get` method.
However, if we were to construct an `OpDescription` from the information stored in the `OpSet::m_dialectOps` object, we'd construct a `VisitorKey`-local object in some helper method for each dialect op in set. The Visitor needs to reference that since it doesn't own the `OpDescription` objects. This will lead to dangling pointers. However, making it static inside this `VisitorKey` method would also not work since all keys would use the same memory location.

Making the VisitorTemplate own the objects is also no appropriate solution in terms of perf / storage, so we'd rather get rid of the OpDescription approach in the `VisitorTemplate`. Happy for opinions.

nhaehnle commented on include/llvm-dialects/Dialect/OpSet.h:
```diff
@@ -0,0 +1,227 @@
+    for (const unsigned CoreOp : m_coreOpcodes) {
+      if (inst.getOpcode() == CoreOp)
+        return true;
+    }
```
For performance, this should make use of the fact that `m_coreOpcodes` is a set ;)

nhaehnle commented on include/llvm-dialects/Dialect/OpSet.h:
```diff
@@ -0,0 +1,227 @@
+        if (containsIntrinsicID(op))
+          return true;
+      }
+    }
```
I still think this should assert that there is only a single opcode / intrinsic ID in the OpDescription.

nhaehnle commented:
This sounds like using an `OpMap` is the way to go.

nhaehnle commented on include/llvm-dialects/Dialect/OpSet.h:
```diff
@@ -0,0 +1,227 @@
+        if (func.getIntrinsicID() == Intrinsic)
+          return true;
+      }
+    }
```
m_intrinsicIDs is a set. And I think this can early-out by returning `false` here.

nhaehnle commented on include/llvm-dialects/Dialect/OpSet.h:
```diff
@@ -0,0 +1,227 @@
+      }
+    }
+
+    return isMatchingDialectOp(desc.getMnemonic());
```
Can't this compare the OpDescription directly?

nhaehnle commented on include/llvm-dialects/Dialect/OpSet.h:
```diff
@@ -0,0 +1,227 @@
+    if (desc.isCoreOp()) {
+      for (const unsigned op : desc.getOpcodes())
+        m_coreOpcodes.insert(op);
+    }
```
Early-out here and below.

nhaehnle commented on include/llvm-dialects/Dialect/OpMap.h:
```diff
@@ -0,0 +1,357 @@
+    if (desc.isCoreOp() || desc.isIntrinsic()) {
+      for (const unsigned op : desc.getOpcodes()) {
+        if (containsCoreOp(op) || containsIntrinsic(op))
+          return true;
```
This needs to only check either `containsCoreOp` or `containsIntrinsic`, depending on the incoming `desc`.

nhaehnle commented on include/llvm-dialects/Dialect/OpMap.h:
```diff
@@ -0,0 +1,357 @@
+    if (auto *CI = dyn_cast<CallInst>(&inst)) {
+      const Function *Callee = CI->getCalledFunction();
+      if (!Callee)
+        return {};
```
This case could still fall through to the core opcode case, since `Instruction::Call` could be in `m_coreOpcodes`.

nhaehnle commented on include/llvm-dialects/Dialect/Dialect.h:
```diff
@@ -245,11 +245,14 @@ class DialectExtensionRegistration {
+bool isSimpleOperation(const llvm::CallInst *i, llvm::StringRef mnemonic);
+bool isOverloadedOperation(const llvm::CallInst *i, llvm::StringRef mnemonic);
+bool isOperationDecl(llvm::StringRef fn, bool isOverloaded,
+                     llvm::StringRef mnemonic);
```
I still think this would be better served with a Function argument for consistency. I think the one place it's used without a function in OpSet could be solved like in OpMap instead.

nhaehnle commented:
Interesting, yes. I think the "key" part of the iterator should just be an OpDescription? That would be consist with the "base" interface for looking things up in the map.

tsymalla commented:
Right, this is the general idea I have (plus some additional state to keep track of what we are actually looking at).
However, this requires to add a default constructor for the `OpDescription`, which we don't want to expose as public API, otherwise we cannot initialize storage in the constructor. We could befriend the `OpMap` to the `OpDescription`, which is ugly as well.
I am not certain it is really that bad we add an default `OpDescription` constructor, since we never do any checks in the remaining `OpDescription` constructors, so we could initialize that with arbitrary (invalid) state as well.

tsymalla commented:
> Right, this is the general idea I have (plus some additional state to keep track of what we are actually looking at). However, this requires to add a default constructor for the `OpDescription`, which we don't want to expose as public API, otherwise we cannot initialize storage in the constructor. We could befriend the `OpMap` to the `OpDescription`, which is ugly as well. I am not certain it is really that bad we add an default `OpDescription` constructor, since we never do any checks in the remaining `OpDescription` constructors, so we could initialize that with arbitrary (invalid) state as well.

(we could derive the iterator from `std::pair<OpDescription, ValueT *>` as well, but that's also ugly)

tsymalla commented:
Added initial refactoring stub using an `OpMapIterator`. This works, but I am quite unhappy with the current solution, so open for suggestions.

First, it is not easy to write an iterator around iterators of two different types. I observed this issue since the beginning. Second, storing an offset into the internal containers seems to work, but it is quite ugly. Storing pointers to the `OpMap` data structures internally seems not to be simply doable.
Third, the current approach uses an O(n) loop for every `operator++`  invocation in the core operations and intrinsic operations state, since it uses `std::advance` on the `DenseMapIterator`. I'd rather pass in an pointer to the previously locally declared iterator and increment that, if it's not `nullptr`, but that requires some ugly workarounds to make it work as well. So, I am leaving that for review at the moment and am very happy for suggestions. Rebase will follow in a sec.

nhaehnle commented:
> Storing iterators to the OpMap data structures internally seems not to be simply doable (since we need to express the constness as defined in the DenseMapIterator in the OpMapIterator as well).

Yeah, being generic over `const`-ness is basically not possible in C++.

The canonical trick here is to define a base CRTP class along the lines of:
```
  template <typename IterT, typename ValueT, typename MapIterator, typename DialectOpIterator>
  struct base_iterator {
  private:
    struct TaggedCoreIterator : MapIterator { /* probably need constructors */ };
    struct TaggedIntrinsicIterator : MapIterator { /* probably need constructors */ };

    const OpMapT *m_map;
    std::variant<DialectOpIterator, CoreIterator, IntrinsicIterator> m_it;

  public:
    ...
  };
```
and then instantiate as
```
  struct const_iterator : base_iterator<iterator, const ValueT, DenseMap<unsigned, ValueT>::const_iterator, SmallVectorImpl<...>::const_iterator> {
    ... will need at least some constructors here ...
  };

  // and then analogously for iterator
```

`operator*` can perhaps just return a `std::pair<OpDescription, ValueT &>` by value. In that case `operator->` has to return a wrapper type.

nhaehnle reviewed (commented):
> Storing iterators to the OpMap data structures internally seems not to be simply doable (since we need to express the constness as defined in the DenseMapIterator in the OpMapIterator as well).

Yeah, being generic over `const`-ness is basically not possible in C++.

The canonical trick here is to define a base CRTP class along the lines of:
```
  template <typename IterT, typename ValueT, typename MapIterator, typename DialectOpIterator>
  struct base_iterator {
  private:
    struct TaggedCoreIterator : MapIterator { /* probably need constructors */ };
    struct TaggedIntrinsicIterator : MapIterator { /* probably need constructors */ };

    const OpMapT *m_map;
    std::variant<DialectOpIterator, CoreIterator, IntrinsicIterator> m_it;

  public:
    ...
  };
```
and then instantiate as
```
  struct const_iterator : base_iterator<iterator, const ValueT, DenseMap<unsigned, ValueT>::const_iterator, SmallVectorImpl<...>::const_iterator> {
    ... will need at least some constructors here ...
  };

  // and then analogously for iterator
```

`operator*` can perhaps just return a `std::pair<OpDescription, ValueT &>` by value. In that case `operator->` has to return a wrapper type.

tsymalla commented:
I am sorry the next update took a little bit longer, this was because of known reasons.

I implemented the const- and non-const version via `std::conditional_t`, which helps integrating everything in a single class, without inheritance.
This required a little bit of fiddling around with the function members from `OpMap`. I ultimatively decided to make the `OpMapIterator` smart, which means to move all relevant logic to generate the right iterator to the `OpMap` internals to the `OpMapIterator`.
Plus, the `Visitor` now should fully support the `OpMap`, and for me, offline, it passes. I see some failures on CI which seem to have to do with repo? Maybe something is wrong, which needs to be fixed in `main` as well - since there has been no commit for quite a while now, we'd test that out.

Performance-wise, this indeed is a bit of overhead, namingly the state-machine approach when iterating through the `OpMap`. I also added a bit of preprocessor "magic" to not have all the cumbersome `iterator` / `const_iterator` overloads inside the `OpMap`...

Happy for comments.

tsymalla commented:
The ubsan+asan Jobs complain:
_30.19 info: A new version of repo is available
30.19 warning: repo is not tracking a remote branch, so it will not receive updates
30.19 repo reset: error: Entry '.flake8' not uptodate. Cannot merge.
30.19 fatal: Could not reset index file to revision 'v2.36.1^0'.
30.19 
------
dialects.Dockerfile:34
--------------------
  32 |     
  33 |     # Sync the repos. Replace the base repo with a freshly checked-out one.
  34 | >>> RUN /vulkandriver/update-dialects.sh
  35 |     
  36 |     # Build llvm-dialects._

Flakebi commented:
The CI needs to be updated to not use the shadercache anymore. I’ll take a look at that.

Flakebi commented:
Should be fixed in #65

tsymalla-AMD commented:
Thanks, @Flakebi !
Rebased this PR.

tsymalla-AMD commented:
Fix issue caught by UBSan

tsymalla-AMD commented:
Remove unused member

tsymalla-AMD commented:
Change: Hide `OpMapIterator` constructors from other clients, since its functionality should only be exposed via the `OpMap`.

tsymalla commented:
Fixed iteration bug when invoking `begin()` with empty core op container.
Added iterator test.

nhaehnle commented on include/llvm-dialects/Dialect/OpMap.h:
```diff
@@ -0,0 +1,814 @@
+    return findOrConstruct(desc);
+  }
+  ValueT *operator[](const Function &func) { return findPointer(func); }
+  ValueT *operator[](const Instruction &inst) { return findPointer(inst); }
```
Remove the `operator[]` overloads that return a pointer. That's too surprising.

nhaehnle commented on include/llvm-dialects/Dialect/OpMap.h:
```diff
@@ -0,0 +1,814 @@
+  // directly return a reference if the key already exists.
+  // -------------------------------------------------------------
+
+  ValueT &findOrConstruct(const OpDescription &desc, const ValueT &val) {
```
This API is too surprising. When you say "find or construct" I hear "either find (and leave the map unmodified) or construct a new element". What this method does instead is to unconditionally assign `val` to `desc`.

It seems this is primarily required for the copy constructor. Could you instead just implement `try_emplace`? That is:
```c++
  template <typename... Ts>
  std::pair<iterator, bool> try_emplace(const OpDescription &desc, Ts &&... args) {
    ...
    auto [it, inserted] = m_coreOpcodes.try_emplace(op, std::forward<Ts>(args)...);
    return {iterator_t(this, it, false), inserted};
    ...
  }
```

tsymalla replied:
Yeah, makes sense. This requires some adjustments. Generally, an issue for the Visitor refactoring is that we constrained the `OpDescriptions` to be inserted to be single-opcodes. This means, something like that:

```
      m_opMap.insert(try_emplace, {handlerIdx});
```

does not neccessarily work.

I'd rather remove that constraint for insertion and return an iterator range instead. Same goes for the `OpMapIterator::erase()` method, since it is rather inefficient to only erase a single operation at a time.

nhaehnle replied:
Hmm, but if the OpDescription has multiple opcodes, then inserted locations don't necessarily fall into a contiguous range of iterators.

For the longer term, I'd rather remove the support for multi-opcodes/multi-intrinsics in OpDescription. It's always been more of a hack.

nhaehnle commented on include/llvm-dialects/Dialect/Visitor.h:
```diff
@@ -287,9 +300,8 @@ class VisitorBase {
-  llvm::DenseMap<unsigned, HandlerRange> m_coreOpcodeMap;
-  llvm::DenseMap<unsigned, HandlerRange> m_intrinsicIdMap;
-  std::vector<std::pair<const OpDescription *, HandlerRange>> m_dialectCases;
+  DenseMap<unsigned, HandlerRange> m_map2;
```
This looks like an leftover from an earlier version?

nhaehnle commented on lib/Dialect/Visitor.cpp:
```diff
@@ -51,37 +51,58 @@ void VisitorTemplate::add(VisitorKey key, VisitorCallback *fn,
-      for (unsigned opcode : description->getOpcodes())
-        m_coreOpcodeMap[opcode].push_back(handlerIdx);
+      for (unsigned op : description->getOpcodes())
+        m_opMap[OpDescription::fromCoreOp(op)].push_back(handlerIdx);
```
It should be possible to simplify this entire block of code down to `m_opMap[*description].push_back(handleIdx);`

nhaehnle commented on lib/Dialect/Visitor.cpp:
```diff
@@ -51,37 +51,58 @@ void VisitorTemplate::add(VisitorKey key, VisitorCallback *fn,
+        SmallVector<unsigned> handlers;
+        handlers.push_back(handlerIdx);
+        m_opMap[desc] = std::move(handlers);
+      }
```
This can just be `m_opMap[desc].push_back(handlerIdx)`

nhaehnle reviewed (commented):
Sorry it took me so long to get around to it. The iterator solution looks pretty good now and I think it's really quite close. Though there are some interface issues that I think should really be addressed before we commit it. As a general rule, the interfaces of OpMap and OpSet should be as unsurprising and as close to the corresponding STL classes where possible. There are some gaps there still, see inline.

tsymalla commented:
Addressed review comments.
Renamed tests.
Implemented try_emplace iterator-based approach.
Fixed issue with new try_emplace-based approach, where the iterator would compare DenseMap iterators built in different epochs.
Added some more tests to demonstrate functionality when trying to mix different kinds of operations in the OpMap.

nhaehnle reviewed (approved):
Thanks! While I haven't gone over it again with a comb, I think the overall design is solid and we can continue to iterate if anything comes up.

tsymalla commented:
Two minor changes: Removed a superfluous code line from `OpSet` and replaced a call to `OpMap::insert` in the Visitor with a call to `OpMap::operator[]` for consistency reasons. Should not make any difference and I'll merge after the checks have passed.
